src/h3lib/lib/h3Index.c

```diff
@@ -129,6 +131,265 @@ int H3_EXPORT(isValidCell)(H3Index h) {
+    | 100 |     100 | 011 | 010      |                  000 | OK      |
+    | 101 |     100 | 010 | 001      |                  000 | OK      |
+    | 110 |     100 | 001 | 000      |                  000 | OK      |
+    | 111 |     100 | 000 | 111*     |                  100 | invalid |
+
+      *: carry happened
+
+
+    Note: only care about identifying the *lowest* 7.
+
+    Examples with multiple digits:
+
+    |    d    | d & MHI |    ~d   | ~d - MLO | d & MHI & (~d - MLO) |  result |
+    |---------|---------|---------|----------|----------------------|---------|
+    | 111.111 | 100.100 | 000.000 | 110.111* |              100.100 | invalid |
+    | 110.111 | 100.100 | 001.000 | 111.111* |              100.100 | invalid |
+    | 110.110 | 100.100 | 001.001 | 000.000  |              000.000 | OK      |
+
+      *: carry happened
+
+    In the second example with 110.111, we "misidentify" the 110 as a 7, due
+    to a carry from the lower bits. But this is OK because we correctly
+    identify the lowest (only, in this example) 7 just before it.
+
+    We only have to worry about carries affecting higher bits in the case of
+    a 7; all other digits (0--6) don't cause a carry when computing ~d - MLO.
+    So even though a 7 can affect the results of higher bits, this is OK
+    because we will always correctly identify the lowest 7.
+
+    For further notes, see the discussion here:
+    https://github.com/uber/h3/pull/496#discussion_r795851046
+
+    */
+    {
+        // does setting static make this faster?
+        const uint64_t MHI = 0b100100100100100100100100100100100100100100100;
+        const uint64_t MLO = MHI >> 2;
+
+        H3Index g = h;
+
+        int shift = 3 * (15 - res);
+        g >>= shift;
+        g <<= shift;
+
+        if (g & MHI & (~g - MLO)) return false;
+    }
+
+    // Check that all unused digits after `res` are set to 7 (INVALID_DIGIT).
+    // Bit shift to avoid looping through digits.
+    if (res < 15) {
+        H3Index g = ~h;
+
+        int shift = 19 + 3 * res;
+        g <<= shift;
+        g >>= shift;
+
+        if (g) return false;
+    }
+
+    // One final validation just for pentagons:
+    // Pentagon cells start with a sequence of 0's (CENTER_DIGIT's).
+    // The first nonzero digit can't be a 1 (i.e., "deleted subsequence",
+    // PENTAGON_SKIPPED_DIGIT, or K_AXES_DIGIT).
+    if (isBaseCellPentagonArr[bc]) {
+        H3Index g = h;
+        g <<= 19;
+        g >>= 19;  // at this point, g < 2^45 - 1
+
+        if (g == 0) return true;  // all zeros (res 15 pentagon)
+
+        // Converting g to a double essentially puts log2(g) into the
+        // exponent bits of f (with an offset).
+        // This works up to about 2^53, which is good enough for our purposes.
+        // (It's 2^53 because doubles have 52 bits in the mantissa.)
+        double f = g;
+
+        // reinterpret the double bits as uint64_t
+        g = *(uint64_t *)&f;
```

Comment:
I'm unclear if this is safe - it may be necessary to use a different implementation on a platform like ARM (we can probably test this on Raspberry Pi). We should check if UBSan complains about this.